File: binding/include/multimedia/libqaudiobufferoutput.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qaudiobufferoutput.html
pub const qaudiobufferoutput = struct {

    /// New constructs a new QAudioBufferOutput object.
    pub fn New() QtC.QAudioBufferOutput {
        return qtc.QAudioBufferOutput_new();
    }


    /// New2 constructs a new QAudioBufferOutput object.
    pub fn New2(format: ?*anyopaque) QtC.QAudioBufferOutput {
        return qtc.QAudioBufferOutput_new2(@ptrCast(format));
    }


    /// New3 constructs a new QAudioBufferOutput object.
    pub fn New3(parent: ?*anyopaque) QtC.QAudioBufferOutput {
        return qtc.QAudioBufferOutput_new3(@ptrCast(parent));
    }


    /// New4 constructs a new QAudioBufferOutput object.
    pub fn New4(format: ?*anyopaque, parent: ?*anyopaque) QtC.QAudioBufferOutput {
        return qtc.QAudioBufferOutput_new4(@ptrCast(format), @ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QAudioBufferOutput_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qaudiobufferoutput.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Format(self: ?*anyopaque, ) QtC.QAudioFormat {
        return qtc.QAudioBufferOutput_Format(@ptrCast(self));
    }

    pub fn AudioBufferReceived(self: ?*anyopaque, buffer: ?*anyopaque) void {
        qtc.QAudioBufferOutput_AudioBufferReceived(@ptrCast(self), @ptrCast(buffer));
    }

    pub fn OnAudioBufferReceived(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QAudioBufferOutput_Connect_AudioBufferReceived(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QAudioBufferOutput_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qaudiobufferoutput.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QAudioBufferOutput_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qaudiobufferoutput.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QAudioBufferOutput_Delete(@ptrCast(self));
    }
};

File: binding/src/multimedia/libqaudiobufferoutput.cpp
#include <QAudioBuffer>
#include <QAudioBufferOutput>
#include <QAudioFormat>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qaudiobufferoutput.h>
#include "libqaudiobufferoutput.h"
#include "libqaudiobufferoutput.hxx"

QAudioBufferOutput* QAudioBufferOutput_new() {
	 return new QAudioBufferOutput();
}

QAudioBufferOutput* QAudioBufferOutput_new2(const QAudioFormat* format) {
	 return new QAudioBufferOutput(*format);
}

QAudioBufferOutput* QAudioBufferOutput_new3(QObject* parent) {
	 return new QAudioBufferOutput(parent);
}

QAudioBufferOutput* QAudioBufferOutput_new4(const QAudioFormat* format, QObject* parent) {
	 return new QAudioBufferOutput(*format, parent);
}

libqt_string QAudioBufferOutput_Tr(const char* s) {
	QString _ret = QAudioBufferOutput::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QAudioFormat* QAudioBufferOutput_Format(const QAudioBufferOutput* self) {
	return new QAudioFormat(self->format());
}

void QAudioBufferOutput_AudioBufferReceived(QAudioBufferOutput* self, const QAudioBuffer* buffer) {
	self->audioBufferReceived(*buffer);
}

void QAudioBufferOutput_Connect_AudioBufferReceived(QAudioBufferOutput* self, intptr_t slot) {
    void (*slotFunc)(QAudioBufferOutput*, const QAudioBuffer*) = reinterpret_cast<void (*)(QAudioBufferOutput*, const QAudioBuffer*)>(slot);
    QAudioBufferOutput::connect(self, &QAudioBufferOutput::audioBufferReceived, [self, slotFunc](const QAudioBuffer& buffer) {
	slotFunc(self, buffer);
    });
}

libqt_string QAudioBufferOutput_Tr2(const char* s, const char* c) {
	QString _ret = QAudioBufferOutput::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QAudioBufferOutput_Tr3(const char* s, const char* c, int n) {
	QString _ret = QAudioBufferOutput::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QAudioBufferOutput_Delete(QAudioBufferOutput* self) {
    delete self;
}

